tinygrad/runtime/ops_gpu.py

```diff
@@ -18,65 +18,72 @@
 
 class _CL:
   def __init__(self):
+    self.events_in_flight = []
+  def post_init(self, device=None):
     platforms: List[List[cl.Device]] = [y for y in ([x.get_devices(device_type=cl.device_type.GPU) for x in cl.get_platforms()] + [x.get_devices(device_type=cl.device_type.CPU) for x in cl.get_platforms()]) if len(y)]
-    devices: List[cl.Device] = [x for x in platforms[getenv('CL_PLATFORM', 0)] if x.name not in getenv('CL_EXCLUDE', '').split(',')]
-    if DEBUG >= 1: print(f"using devices: {[d.hashable_model_and_version_identifier for d in devices]}")
-    self.cl_ctx: cl.Context = cl.Context(devices=devices)
-    self.cl_queue: List[cl.CommandQueue] = [cl.CommandQueue(self.cl_ctx, device=device, properties=cl.command_queue_properties.PROFILING_ENABLE) for device in self.cl_ctx.devices]
-
+    self.cl_platform = cl.get_platforms()[getenv('CL_PLATFORM', 0)]
+    self.cl_ctxs: List[cl.Context] = [cl.Context(devices=[x]) for x in platforms[getenv('CL_PLATFORM', 0)] if x.name not in getenv('CL_EXCLUDE', "").split(",")] if device is None else [cl.Context(devices=[platforms[getenv('CL_PLATFORM', 0)][device]])]
```

Comment:
What's the advantage of a different context for each device?

Reply:
P2P on opencl requires that the devices be in different contexts, but I found that getting multiple contexts on the same device from multiple processes causes the queues to hang.